Add vc-got-command and use it instead of vc-got--call

zmyrgel opened the pull request:
vc-command has lot of niceties compared to just running got directly.
Provide vc-got-command wrapper like other vc backends and use it to run most of the got commands. Some commands left to use vc-got--call as otherwise they otherwise they show excessive amount of info in echo area and *messages*.

zmyrgel committed:
add vc-got-command as copy of vc-command

this is similar as is done with other vc backends, helps run synchronous commands.

zmyrgel committed:
rewrite most of vc-got-call uses to use vc-got-command

benefit of this is better file argument handling. Some commands left to
use vc-got--call as using vc-got-command was too "chatty".

while at it fix or add some comments.

zmyrgel committed:
fix unbalanced parens

zmyrgel committed:
restore accidently deleted delete-matching-lines call

omar-polo commented:
I didn't know about `vc-do-command`!  I'm going to test the changes but I like the diff, thank you!

omar-polo commented:
I'll need to better understand what's going on, but this at least broke committing since apparently somehow now the commit message string is passed with text properties.

```
Debugger entered--Lisp error: (wrong-type-argument characterp #("foo\n\n" 0 3 (jit-lock-defer-multiline t face log-edit-summary fontified t field header)))
  vc-delistify(("commit" "-m" (#("foo\n\n" 0 3 (jit-lock-defer-multiline t face log-edit-summary fontified t field header)))))
  vc-do-command(t 0 "got" ("/home/op/w/telescope/telescope.h") "commit" "-m" (#("foo\n\n" 0 3 (jit-lock-defer-multiline t face log-edit-summary fontified t field header))))
  vc-got-command(t 0 ("/home/op/w/telescope/telescope.h") "commit" "-m" (#("foo\n\n" 0 3 (jit-lock-defer-multiline t face log-edit-summary fontified t field header))))
  vc-got-checkin(("/home/op/w/telescope/telescope.h") #("Summary: foo\n" 0 8 (jit-lock-defer-multiline t face log-edit-header fontified t field header) 8 9 (jit-lock-defer-multiline t fontified t field header) 9 12 (jit-lock-defer-multiline t face log-edit-summary fontified t field header) 12 13 (fontified t)) nil)
  vc-call-backend(Got checkin ("/home/op/w/telescope/telescope.h") #("Summary: foo\n" 0 8 (jit-lock-defer-multiline t face log-edit-header fontified t field header) 8 9 (jit-lock-defer-multiline t fontified t field header) 9 12 (jit-lock-defer-multiline t face log-edit-summary fontified t field header) 12 13 (fontified t)) nil)
```

omar-polo commented on vc-got.el:
```diff
@@ -413,20 +415,19 @@ files on disk."
 
 (defun vc-got--branch (name)
   "Try to create and switch to the branch called NAME."
+  ;; TODO: does branch change files or does it require update
```
IIRC this function is supposed to just create a new branch, not switch to an existing one, so it does not require an update.  I think it's expected to fail if the target branch already exists.

zmyrgel replied:
Yeah, I was thinking does it need "  (let (process-file-side-effects)" at the start

omar-polo replied:
That's a good question.  Technically this changes some files in `.got` and in the bare repository, but I doubt this is relevant (noone has these files open in a emacs buffer, and it's a kind of implementation detail).  So yeah, it would make sense to me to add `(let (process-file-side-effects)` here.

zmyrgel committed:
fix buffer parameters in vc-got-command calls

no need to use temp buffers in most of commands, pass in `nil' to use
shared *vc* buffer.

zmyrgel commented:
Added some fixes to vc-got-command buffer parameters. At least diff worked but committing is still broken. Will look into it next.

omar-polo committed:
remove copy/pasted block

omar-polo committed:
fix vc-got-checkin

omar-polo commented:
I still have to properly test it, but that seemed to have done the trick.  Will continue use it and see if anything else is broken :)

omar-polo commented:
There's a similar issue in `vc-got--tag-callback` now, but i'm not brave enough to commit this xD

```
--- vc-got.el
+++ vc-got.el
@@ -839,11 +839,13 @@ Creates the TAG using the content of the current buffe
   (interactive)
   (let ((msg (buffer-substring-no-properties (point-min)
                                              (point-max))))
-    (vc-got-command nil 0 nil "tag"
-                    "-m"
-                    (log-edit-extract-headers nil msg)
-                    "--"
-                    tag)))
+    (apply #'vc-got-command nil 0 nil "tag"
+			 "-m"
+			 (flatten-list
+			  (list
+			   (log-edit-extract-headers nil msg)
+			   "--"
+			   tag)))))
 
 (defun vc-got-create-tag (_dir name branchp)
   "Attach the tag NAME to the state of the worktree.
```

zmyrgel commented:
> There's a similar issue in `vc-got--tag-callback` now, but i'm not brave enough to commit this xD
> 
> ```
> --- vc-got.el
> +++ vc-got.el
> @@ -839,11 +839,13 @@ Creates the TAG using the content of the current buffe
>    (interactive)
>    (let ((msg (buffer-substring-no-properties (point-min)
>                                               (point-max))))
> -    (vc-got-command nil 0 nil "tag"
> -                    "-m"
> -                    (log-edit-extract-headers nil msg)
> -                    "--"
> -                    tag)))
> +    (apply #'vc-got-command nil 0 nil "tag"
> +			 "-m"
> +			 (flatten-list
> +			  (list
> +			   (log-edit-extract-headers nil msg)
> +			   "--"
> +			   tag)))))
>  
>  (defun vc-got-create-tag (_dir name branchp)
>    "Attach the tag NAME to the state of the worktree.
> ```

I think that looks good as well.

zmyrgel committed:
fix vc-got-command rest arguments

use apply when there is variable amount of parameters.

zmyrgel commented:
Actually, looking at it more closely I think the (append (list ..) (list ..)) would be better syntax for it.

I iterated the vc-got-commands, there were quite a lot of places that might take list values so updated those to use apply.
Still getting errors from mode-lines:
`concat: Args out of range: #("-0" 0 2 (face vc-up-to-date-state help-echo "Up to date file under the Got version control system")), 0, 4`

zmyrgel committed:
fix vc-got-mode-line to not refer past length of def

zmyrgel commented:
And found a fix for the mode-line. I think that fixes my known issues with the PR but I'll spend some time testing it further.

omar-polo commented:
Oh good catch! I wonder why I haven't hit that before.  I'll test as well tomorrow :)

thank you!

zmyrgel committed:
fix deletion of separator lines in log buffers

zmyrgel committed:
fix vc-got--info command return value use

The got info command uses return value of 1 when executed for files
which are not part of got work tree which caused the emacs to show those
as errors. To avoid these errors when accessing non-registered file, tune up the
vc-got--info function to ignore the error cases when executing the
vc-got-command and only check the command return value in
vc-got-working-revision.

zmyrgel committed:
make vc-got--update asynchronous

Make the update command run asynchronously as updates in large directory
trees will otherwise lock emacs for a long while.

zmyrgel committed:
fix docstring for vc-got--proc-filter

zmyrgel committed:
use shorter way to poll branch from user in vc-got-merge-branch

zmyrgel commented:
I noticed a small issue with "got info" which I fixed. I also added async version for the vc-update to avoid "stuck emacs" for the update duration. It takes a while to wait for /usr/ports update to complete.

zmyrgel commented:
@omar-polo ping

I haven't found any regressions and I have other commits pending which I don't wish to push here.

omar-polo commented:
Apologize, I forgot about this PR.  Looks good to me, let's merge this.

Thanks!
